File: Download/app/src/main/java/com/cyq/download/MainActivity.java
package com.cyq.download;

import android.Manifest;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.tbruyelle.rxpermissions2.RxPermissions;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private ProgressBar mProgress;
    private TextView mPresent;
    private Button mBtnDownload;
    private Button mBtnCancel;
    private RxPermissions rxPermissions;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initView();
        initPermissions();
    }

    private void initView() {
        mProgress = findViewById(R.id.progress);
        mPresent = findViewById(R.id.tv_progress);
        mBtnCancel = findViewById(R.id.btn_cancel);
        mBtnDownload = findViewById(R.id.btn_download);
        mBtnDownload.setOnClickListener(this);
        mBtnCancel.setOnClickListener(this);
    }

    /**
     * 申请文件读写权限
     */
    private void initPermissions() {
        rxPermissions = new RxPermissions(this);
        rxPermissions.request(Manifest.permission.WRITE_EXTERNAL_STORAGE,
                Manifest.permission.READ_EXTERNAL_STORAGE)
                .subscribe();
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_download:

                break;
            case R.id.btn_cancel:

                break;
            default:
        }
    }
}

File: ConstraintLayoutDemo/network/src/main/java/com/example/network/MainActivity.kt
package com.example.network

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.network.bean.BannerBean
import com.example.network.bean.ChildBannerBean
import com.example.network.http.ApiException
import com.example.network.http.BaseRsp
import com.example.network.http.HttpUtil
import com.example.network.http.MyObserver
import com.google.gson.Gson
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.Scheduler
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    private var mTvResult: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mTvResult = findViewById(R.id.tv_result)

        findViewById<Button>(R.id.btn_get).setOnClickListener {
            HttpUtil.createService(RequestApi::class.java)
                .getBanner()
                .enqueue(object : Callback<BannerBean> {
                    override fun onResponse(
                        call: Call<BannerBean>,
                        response: Response<BannerBean>
                    ) {
                        Log.d(
                            "test",
                            "请求成功：${Gson().toJson(response.body())}  ${Thread.currentThread()}"
                        )
                        mTvResult?.text = Gson().toJson(response.body())
                    }

                    override fun onFailure(call: Call<BannerBean>, t: Throwable) {
                        Log.d("test", "请求失败：$call")
                    }
                })
        }

        findViewById<Button>(R.id.btn_rxjava_get).setOnClickListener {
//            HttpUtil.createService(RequestApi::class.java)
//                .getBanner2()
//                .subscribeOn(Schedulers.io())
//                .observeOn(AndroidSchedulers.mainThread())
//                .subscribe(object : MyObserver<BannerBean>(this) {
//                    override fun onSuccess(t: BannerBean) {
//                        Log.d("test", "onNext---->$t")
//                        mTvResult?.text = Gson().toJson(t)
//                    }
//
//                    override fun onFail(apiException: ApiException) {
//                        Log.d("test", "onError---->$apiException")
//                        apiException.printStackTrace()
//                    }
//                })
            HttpUtil.createService(RequestApi::class.java)
                .getBanner3()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                    object : MyObserver<BaseRsp<List<ChildBannerBean>>>(this) {
                        override fun onSuccess(t: BaseRsp<List<ChildBannerBean>>) {
                            if (t.isSuccess()) {
                                Log.d("test", "请求成功：${t.data}")
                            } else {
                                Log.d("test", "请求失败：${t.errorMsg}")
                            }
                        }

                        override fun onFail(apiException: ApiException) {
                            Log.d("test", "onError---->$apiException")
                        }
                    }
                )
        }
    }
}